[stdlib] Reintroduce _getBool for ABI

Azoy opened the pull request:
https://github.com/apple/swift/pull/20948 didn't make the 5.0 branch. Reintroduce the API function as internal so that older binaries can still call it. As requested from Ben: https://github.com/apple/swift/pull/21872#issuecomment-464581970

cc: @airspeedswift

Azoy committed:
Reintroduce _getBool for ABI

move to legacy

slavapestov commented on stdlib/public/core/Bool.swift:
```diff
@@ -170,6 +170,13 @@ extension Bool : _ExpressibleByBuiltinBooleanLiteral, ExpressibleByBooleanLitera
 
+// This is an old intrinsic known by the compiler, but exists solely for ABI
+// compatibility at this point.
+@usableFromInline @_transparent
```
Can you remove the @_transparent as well?

@airspeedswift Do you know of an example where a user binary was calling this, or is it purely defensive?

Azoy replied:
Off the top of my head, I know the compiler would call this for `IsExpr`s.
```swift
let x = 3 is Int
```

slavapestov replied:
Right, I would expect any call emitted by the compiler would directly apply the function, so it would be inlined away. A user would have to explicitly write a call to _getBool() and make use of the result, which is non-trivial since it is a built-in type. But perhaps we had a bug where it wasn't inlined away somewhere?

airspeedswift replied:
Purely defensive. Like you say, if there's no way for a user to call it. If you're certain there's no way the compiler could emit this symbol without just inlining the implementation, we can leave it removed, but it'd be a shame if we're wrong so the reasoning needs to be solid. WDYT?

benrimmington replied:
Could `_getBool(_:)` be moved to the [LegacyABI.swift](https://github.com/apple/swift/blob/master/stdlib/public/core/LegacyABI.swift) file?

And possibly use one of the following:

- `@available(swift, obsoleted: 5.0)`
- `@available(*, unavailable, renamed: "Bool.init(_:)")`
- `@available(*, unavailable, renamed: "Bool.init(_builtinBooleanLiteral:)")`

Azoy commented:
Closing in favor of: https://github.com/apple/swift/pull/23753

airspeedswift commented:
Oh sorry @Azoy I missed this!

airspeedswift commented:
Let's keep yours since it has the discussion.

airspeedswift commented:
@swift-ci please test

Azoy commented:
@airspeedswift I've moved this to LegacyABI and updated the attributes. Would love to hear what Slava thinks about this, but it should be ready to merge if you guys decide to be defensive here.

airspeedswift commented:
@swift-ci please test

CodaFi commented:
I think we got away with this one.

Let's all just whistle nonchalantly and part ways on this.

Thanks @azoy!
